manuals/wtcdbexport.1
.Dd June 25, 2026
.Dt WTCDBEXPORT 1
.Os
.Sh NAME
.Nm wtcdbexport
.Nd exports items stored in a Windows Explorer thumbnail cache database \
(thumbcache.db) file
.Sh SYNOPSIS
.Nm wtcdbexport
.Op Fl l Ar logfile
.Op Fl t Ar target
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm wtcdbexport
is a utility to export items stored in a Windows Explorer thumbnail cache \
database (thumbcache.db) file
.Pp
.Nm wtcdbexport
is part of the
.Nm libwtcdb
package.
.Nm libwtcdb
is a library to access the Windows Explorer thumbnail cache database (WTCDB) \
format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl l Ar logfile
specify the file in which to log information about the exported items
.It Fl t Ar target
specify the target directory to export to (default is the source filename \
followed by .export)
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# wtcdbexport thumbcache_32.db
wtcdbexport 20260625
.sp
Opening file.
Export completed.
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr wtcdbinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libwtcdb/issues
.Sh COPYRIGHT
Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* expose index and cache entry in API
* libwtcdb
* wtcdbinfo
  - add info handle
* wtcdbexport
  - move file into export handle
* update man pages

* debug print cache type
* debug print index entry flags

* wtcdbexport export items in directories using the entry hash as filename

20101023
* see `git log' for more recent change log
* created initial version based on libnk2 for the library structure and common code
* worked on cache file format support



wtcdbtools/wtcdbexport.c
/*
 * Extracts items from a Windows Explorer thumbnail cache database file
 *
 * Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <narrow_string.h>
#include <system_string.h>
#include <types.h>
#include <wide_string.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "export_handle.h"
#include "log_handle.h"
#include "wtcdbtools_getopt.h"
#include "wtcdbtools_libcerror.h"
#include "wtcdbtools_libclocale.h"
#include "wtcdbtools_libcnotify.h"
#include "wtcdbtools_libcpath.h"
#include "wtcdbtools_libwtcdb.h"
#include "wtcdbtools_output.h"
#include "wtcdbtools_signal.h"
#include "wtcdbtools_unused.h"

export_handle_t *wtcdbexport_export_handle = NULL;
int wtcdbexport_abort                      = 0;

/* Signal handler for wtcdbexport
 */
void wtcdbexport_signal_handler(
      wtcdbtools_signal_t signal WTCDBTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "wtcdbexport_signal_handler";

	WTCDBTOOLS_UNREFERENCED_PARAMETER( signal )

	wtcdbexport_abort = 1;

	if( wtcdbexport_export_handle != NULL )
	{
		if( export_handle_signal_abort(
		     wtcdbexport_export_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal export handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description    = \
		"Use wtcdbexport to export items stored in a Windows Explorer thumbnail cache database (thumbcache.db) file.";

	wtcdbtools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'l', "logfile", "logs information about the exported items" },
		{ 't', "target", "specify the target directory to export to (default is the source filename followed by .export)" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libcerror_error_t *error               = NULL;
	log_handle_t *log_handle               = NULL;
	system_character_t *log_filename       = NULL;
	system_character_t *option_target_path = NULL;
	system_character_t *path_separator     = NULL;
	system_character_t *source             = NULL;
	const char *program                    = "wtcdbexport";
	system_integer_t option                = 0;
	size_t source_length                   = 0;
	int number_of_options                  = (int) ( sizeof( options ) / sizeof( wtcdbtools_option_t ) );
	int result                             = 0;
	int verbose                            = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "wtcdbtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( wtcdbtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	wtcdbtools_output_version_fprint(
	 stdout,
	 program );

	if( wtcdbtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = wtcdbtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				wtcdbtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				wtcdbtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'l':
				log_filename = optarg;

				break;

			case (system_integer_t) 't':
				option_target_path = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				wtcdbtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		wtcdbtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	if( option_target_path == NULL )
	{
		source_length = system_string_length(
		                 source );

		path_separator = system_string_search_character_reverse(
		                  source,
		                  (system_character_t) LIBCPATH_SEPARATOR,
		                  source_length );

		if( path_separator == NULL )
		{
			path_separator = source;
		}
		else
		{
			path_separator++;
		}
		option_target_path = path_separator;
	}
	libcnotify_verbose_set(
	 verbose );
	libwtcdb_notify_set_stream(
	 stderr,
	 NULL );
	libwtcdb_notify_set_verbose(
	 verbose );

	if( log_handle_initialize(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize log handle.\n" );

		goto on_error;
	}
	if( export_handle_initialize(
	     &wtcdbexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize export handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( wtcdbexport_export_handle != NULL );
#endif
	if( export_handle_set_target_path(
	     wtcdbexport_export_handle,
	     option_target_path,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set target path.\n" );

		goto on_error;
	}
	result = export_handle_create_items_export_path(
	          wtcdbexport_export_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to create items export path.\n" );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "%" PRIs_SYSTEM " already exists.\n",
		 wtcdbexport_export_handle->items_export_path );

		goto on_error;
	}
	if( log_handle_open(
	     log_handle,
	     log_filename,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open log file: %" PRIs_SYSTEM ".\n",
		 log_filename );

		goto on_error;
	}
	fprintf(
	 stdout,
	 "Opening file.\n" );

#ifdef TODO_SIGNAL_ABORT
	if( wtcdbtools_signal_attach(
	     wtcdbexport_signal_handler,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to attach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#endif
	if( export_handle_open(
	     wtcdbexport_export_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file.\n" );

		goto on_error;
	}
/* TODO
	fprintf(
	 stdout,
	 "Exporting aliases.\n" );
*/
	result = export_handle_export_file(
	          wtcdbexport_export_handle,
	          log_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to export file.\n" );

		goto on_error;
	}
#ifdef TODO_SIGNAL_ABORT
	if( wtcdbtools_signal_detach(
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to detach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#endif
	if( export_handle_close(
	     wtcdbexport_export_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close export handle.\n" );

		goto on_error;
	}
	if( export_handle_free(
	     &wtcdbexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free export handle.\n" );

		goto on_error;
	}
	if( log_handle_close(
	     log_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close log file.\n" );

		goto on_error;
	}
	if( log_handle_free(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free log handle.\n" );

		goto on_error;
	}
	if( result == 0 )
	{
		fprintf(
		 stdout,
		 "No aliases to export.\n" );
	}
	else
	{
		fprintf(
		 stdout,
		 "Export completed.\n" );
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( wtcdbexport_export_handle != NULL )
	{
		export_handle_free(
		 &wtcdbexport_export_handle,
		 NULL );
	}
	if( log_handle != NULL )
	{
		log_handle_free(
		 &log_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

